### PR TITLE
Excise bound push pt 2

### LINKED ISSUE #243: Resolve scaling and bounds push issues
I've collected some thoughts below from #237. While the perturbation helper in #242 will help WaterTAP developers identify when this affects their models, we should decide a standard way forward for WaterTAP models.

# Problem Statement
WaterTAP models often have small (`< 1e-6`) lower bounds and no or much larger upper bounds. The bounds push in IPOPT, [described in the paper on p. 20](http://cepac.cheme.cmu.edu/pasilectures/biegler/ipopt.pdf), will push small initial values on variable with a small LB no or large UB to ~`1e-2` with default settings, sometimes destroying a "good" initialization. In tests I observed that pushing a small variable up a single order of magnitude (e.g., from `1e-6` to `1e-5`)  would cause issues for the RO unit model.

# Potential Solutions

## 1. Selectively apply `bounds_push` when needed
This is essentially our current strategy. With #242 it may become easier to implement, but could become an issue eventually for end-users.

## 2. Change default IPOPT options when `watertap` is imported
We could update `idaes.cfg["ipopt"]["options"]["bound_push"]` when WaterTAP is imported. A value of `1e-08` is needed to initialize the 1DRO model, but a larger global value could be chosen. We can effectively disable IPOPT's bound's pushing by setting `bound_push` small enough, which ensures a model initialized through standard IDAES methods will not have its starting point modified by IPOPT.
#### Potential Pitfalls
Ensuring a strictly feasible primal starting point is necessary for IPOPT's algorithm to work correctly. Not starting at a strict interior point could negatively affect solver performance.

## 3. Write a custom bounds pusher for WaterTAP models
To ensure we start at a strict interior point, we could write a custom bounds pusher for WaterTAP models which is more aware of small values, and then disable IPOPT's bounds pushing.
#### Potential Pitfalls
It's not clear how to incorporate this into every solve without either monkey-patching Pyomo/IDAES when `watertap` is imported or requiring an additional function call before solving a model. Perhaps we could provide our own "solver" by copying the IDAES `get_solver` strategy.

## 4. Be more careful with bounds
- Ensure our models have variable with bounds that are within an order of magnitude
- Variables are unbounded if bounds are not really necessary

## 5. Scaling small values
- Manually scaling some problematic variables in WaterTAP models to avoid small bounds
- Maintaining `user-scaling`.

### PR DESCRIPTION
**Pull request by bknueven**
## Fixes #243
Part 3/3. Follow-up from #270 and #282.

## Summary/Motivation:
Finish removing `bound_push` and manual Jacobian scaling from the codebase

## Changes proposed in this PR:
- Replace `solve_with_user_scaling` in full_treatment_train folder with `solve_block`
- Update chemistry examples in full_treatment_train folder to align with those in the examples folder
- Tighten default `watertap-ipopt` options numerical tolerances. We were often "getting lucky" with the IDAES default `tol` of `1e-06`.
   -  `tol` to Ipopt default `1e-08` (_scaled_ NLP error)
   - `constr_viol_tol` to `1e-08` (_unscaled_ feasibility tolerance, Ipopt default `1e-04`) 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bknueven** [comment]
Converting this to draft for now: #284 should go first, and will create merge conflicts.

File: watertap/examples/chemistry/tests/test_pure_water_pH.py
Comment (aladshaw3) [inline watertap/examples/chemistry/tests/test_pure_water_pH.py]: Was it necessary to manually change tolerance for this test?

File: watertap/examples/flowsheets/full_treatment_train/chemical_flowsheet_util.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/chemical_flowsheet_util.py:69]: Glad to see that this will not be necessary anymore.

File: watertap/examples/flowsheets/full_treatment_train/electrolyte_scaling_utils.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/electrolyte_scaling_utils.py]: Much of this is removed already in the pending PR #284 (there will be many conflicts after 284 is merged)

File: watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/PostTreatment_SimpleNaOCl_Chlorination.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/PostTreatment_SimpleNaOCl_Chlorination.py]: This file (PostTreatment_SimpleNaOCl_Chlorination.py) is removed by PR #284 (redundant and unused in full flowsheet)

File: watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/PreTreatment_Simple_Softening.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/PreTreatment_Simple_Softening.py]: This file (PreTreatment_Simple_Softening.py) is removed by PR #284 (redundant and unused in full flowsheet)

File: watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/SepRO_plus_Chlorination.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/SepRO_plus_Chlorination.py]: This file (SepRO_plus_Chlorination.py) is removed by PR #284 (redundant and unused in full flowsheet)

File: watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/ZeroDRO_plus_Chlorination.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/ZeroDRO_plus_Chlorination.py]: This file (ZeroDRO_plus_Chlorination.py) is removed by PR #284 (redundant and unused in full flowsheet)

**Review by aladshaw3 — COMMENTED** [review]
I will likely need to review this again because you will have a number of conflicts to resolve after PR #284. Everything looks good to me (as far as the chemistry stuff is concerned). Just a few comments/questions below.

File: watertap/examples/chemistry/tests/test_pure_water_pH.py
Comment (bknueven) [reply watertap/examples/chemistry/tests/test_pure_water_pH.py]: It was, but I think I've found a global set of Ipopt options that work.

This one was strange -- a `tol` of `1e-06` "works" (the tests pass). However, smaller values (e.g., `1e-07`, `1e-09`, would cause regression test failures. Looking at the log, Ipopt happens to get lucky and with `1e-06` solves to a tighter tolerance than `1e-09` due to when the solver chooses to kick out of restoration.

**Comment by bknueven** [comment]
@andrewlee94 would you might having a look at the single failing test to see if the baseline should just be updated or if the difference is significant?

**Comment by aladshaw3** [comment]
@bknueven 

I think @adam-a-a or @TimBartholomew should be the ones to have a look at that assertion error, since it is in the NF model result.

**Comment by adam-a-a** [comment]
> @bknueven
> 
> I think @adam-a-a or @TimBartholomew should be the ones to have a look at that assertion error, since it is in the NF model result.

Actually, @andrewlee94 would be best suited to check this out since the assertion error is attributed to one of the tests for the new zero-order (WT3) NF model; the error has to do with a mismatch in the captured output after calling the report() method.

**Comment by andrewlee94** [comment]
It looks like changing the solver options changed to results of the zero-order models too, and significantly enough that the printed output in the `report` method changed. Seeing as the `report` method was looking for an expected string output (hard-coded in the test), this caused the failure.

The fix is to either update the hard coded string, or to write up something that will construct the expected string from the model output for testing.

**Comment by bknueven** [comment]
Thanks @andrewlee94. After looking in the other tests, you compare the changed reported value to zero. With the new solver options, the report now shows this value as `1.0164e-20`, whereas before it was `9.9000e-07`.

File: watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/posttreatment_ideal_naocl_chlorination_block.py
Comment (adam-a-a) [inline watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/posttreatment_ideal_naocl_chlorination_block.py:699]: Why omit scaling of the mixer?

Comment (adam-a-a) [reply watertap/examples/flowsheets/full_treatment_train/flowsheet_components/chemistry/posttreatment_ideal_naocl_chlorination_block.py:699]: Ah, because it just uses `iscale.constraint_autoscale_large_jac()` which is being eliminated from all code in this PR.

File: watertap/examples/flowsheets/full_treatment_train/flowsheet_components/examples/full_example.py
Comment (adam-a-a) [inline watertap/examples/flowsheets/full_treatment_train/flowsheet_components/examples/full_example.py:208]: commented code

File: watertap/examples/flowsheets/full_treatment_train/util.py
Comment (adam-a-a) [inline watertap/examples/flowsheets/full_treatment_train/util.py]: Can you remind me which tolerance is being adjusted here? I'm guessing solver convergence failed without this tolerance value but unsure why.

File: watertap/examples/flowsheets/full_treatment_train/util.py
Comment (adam-a-a) [inline watertap/examples/flowsheets/full_treatment_train/util.py:31]: Do we need this if we already have the same or similar method in `core/util/initialization` ? Guessing we're keeping this to avoid having to replace imports everywhere else.

File: watertap/examples/flowsheets/full_treatment_train/util.py
Comment (adam-a-a) [inline watertap/examples/flowsheets/full_treatment_train/util.py:35]: Same as above - why not just use check_solve from `core/util/initialization`?

**Review by adam-a-a — DISMISSED** [review]
LGTM- I had some minor comments that can be addressed but shouldn't hold up the merge.

File: watertap/examples/flowsheets/full_treatment_train/util.py
Comment (bknueven) [reply watertap/examples/flowsheets/full_treatment_train/util.py]: It shouldn't be -- that is a carry-over from the tortured history of this branch. I probably need to do a full re-review myself to ensure nothing else like this slipped through.

Comment (bknueven) [inline watertap/examples/flowsheets/full_treatment_train/util.py]:
```suggestion
        solver = get_solver()
```

File: watertap/examples/flowsheets/full_treatment_train/util.py
Comment (aladshaw3) [inline watertap/examples/flowsheets/full_treatment_train/util.py]: One of the chemistry post-treatment blocks is failing a solution check after the tolerance was changed. The model still reaches an optimal solution, but the value is very slightly off now. The failure in the solution check is probably a negligible difference between the values, but I will need to go into it deeper to be certain.

Comment (adam-a-a) [reply watertap/examples/flowsheets/full_treatment_train/util.py]: Looking at the documentation for IPOPT options, I see the default for this convergence tolerance is 1e-8, so I suppose the 1e-11 tolerance yielded a solution with smaller NLP error. As long as we still get "optimal termination" and the solutions are negligibly different, I'd think we'd be better off sticking with the default value of 1e-8.

Comment (bknueven) [reply watertap/examples/flowsheets/full_treatment_train/util.py]: I think we should find the value that will give us reliable results for the chemistry models. By default IDAES relaxes this tolerance to `1e-06`, because that makes the most sense for the typical power plant model. We can, and in my opinion, should make a similar choice.

I'll be working on this today to see if there's a set of global options that works across our WaterTAP models.

Comment (andrewlee94) [reply watertap/examples/flowsheets/full_treatment_train/util.py]: I think there should be a way to overwrite the IDAES default scaling for WaterTap (i.e. set out own set of defaults). John Eslick should know how to do this.

File: watertap/core/plugins/solvers.py
Comment (aladshaw3) [inline watertap/core/plugins/solvers.py]: Same test is failing for the same reason, but the rate of convergence has improved with these changes.  

However, there may be a new problem with this test now: The solver is producing different solutions depending on the operating system now. In windows, it reports a pH of 8.47 (which gives ion distribution of 11.7 % to HOCl and 88.3 % to OCl-), but on Linux it reports a pH of 8.76 (which gives ion distribution of 7.1 % to HOCl and 92.9 % to OCl-). 

To me, this is an unacceptable deviation. At the very least, each OS/build should give the same results.

Comment (aladshaw3) [reply watertap/core/plugins/solvers.py]: Also, the rate of convergence for this test on this branch is significantly worse than the rate of convergence on main (as it currently stands).

**Comment by bknueven** [comment]
@aladshaw3 would you mind having another look at this?

**Review by aladshaw3 — APPROVED** [review]
LGTM

**Review by adam-a-a — APPROVED** [review]
LGTM

**Pull request merged**
